### PR TITLE
feat: support 2fa

### PR DESCRIPTION
**Pull request by sh1luo**
Fix: #81 
Signed-off-by: sh1luo <690898835@qq.com>

### CODE REVIEW

**Comment by sh1luo** [comment]
demo link: http://testb.rescld.cn/

**Comment by hsluoyz** [comment]
@Kininaru plz review.

**Review by Kininaru — COMMENTED** [review]
Code looks good.

But I noticed that you added an API `/api/verify-email-or-phone` as the final step of the application 2FA.

Now, we have `EmailCode` and `PhoneCode` in `RequestForm`, and I think we could reuse the login API: just add an other member variable `WayOf2FA` to `RequestFrom`, and check 2FA in `HandleLoggedIn`.

What's your opinion? 
@hsluoyz @1340908470 @sh1luo @oranges-eating

File: web/src/VerifyModal.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  const handleOk = () => {
+    if (code === "") {
+      Setting.showMessage("error", i18next.t("code:Empty Code"));
+      return;
+    }
+    setConfirmLoading(true);
+    UserBackend.verifyEmailOrPhone(addr, destType, code, userId).then(res => {
+      if (res.status === "ok") {
+        Util.showMessage("success", `Verified successfully`);
```
[the remark this answers is not in the document]
Comment (sh1luo) [reply web/src/VerifyModal.js:40]: I think you are right, and will change it later.

[the remark this answers is not in the document]
Comment (hsluoyz) [reply web/src/VerifyModal.js:40]: Yes. We should re-use existing API as much as possible.

[the remark this answers is not in the document]
Comment (sh1luo) [reply web/src/VerifyModal.js]: This time I re-use the codes in  `controllers/auth.go` and find some bugs during coding within it, then fix them.

Now, everything is ok!

[the remark this answers is not in the document]
Comment (hsluoyz) [reply web/src/VerifyModal.js]: @Kininaru plz review

**Review by Kininaru — APPROVED** [review]

**Comment by hsluoyz** [comment]
@sh1luo plz resolve conflicts.

**Comment by sh1luo** [comment]
> @sh1luo plz resolve conflicts.

ok

**Comment by sh1luo** [comment]
@hsluoyz ok, done. current demo link: http://testb.rescld.cn/

**Comment by hsluoyz** [comment]
@sh1luo resolve conflicts:

image

**Comment by sh1luo** [comment]
I'm solving this problem, it's a little tricky

**Comment by sh1luo** [comment]
> @sh1luo resolve conflicts:
> 
> image

ok, done.

**Comment by hsluoyz** [comment]
@sh1luo link committer to github:

image

**Comment by sh1luo** [comment]
@hsluoyz what's wrong with this PR?

File: controllers/account.go [hunk 2]
```diff
@@ -48,6 +48,7 @@ type RequestForm struct {
 	State       string `json:"state"`
 	RedirectUri string `json:"redirectUri"`
 	Method      string `json:"method"`
+	WayOf2FA    string `json:"wayOf2FA"`
```
Comment (hsluoyz) [inline controllers/account.go:51]: Use `MfaMethod`, put after `Affiliation`

File: controllers/auth.go [hunk 3]
```diff
@@ -38,6 +38,20 @@ func (c *ApiController) HandleLoggedIn(application *object.Application, user *ob
 	userId := user.GetId()
 	resp := &Response{}
 	if form.Type == ResponseTypeLogin {
+		if application.Enable2FA {
```
Comment (hsluoyz) [inline controllers/auth.go:41]: EnableMfa

File: controllers/auth.go [hunk 4]
```diff
@@ -38,6 +38,20 @@ func (c *ApiController) HandleLoggedIn(application *object.Application, user *ob
...
+			util.LogInfo(c.Ctx, "API: [%s] need to verify email/phone", userId)
+			addr := ""
+			if application.WayOf2FA == "Email" {
+				addr = user.Email
+			} else if application.WayOf2FA == "Phone" {
+				addr = user.Phone
+			} else {
+				resp = &Response{Status: "error", Msg: fmt.Sprint("Unknown 2FA Way"), Data: application.WayOf2FA}
```
Comment (hsluoyz) [inline controllers/auth.go:49]: No the word `way` anywhere.

File: web/src/locales/en/data.json [hunk 5]
```diff
@@ -133,7 +134,8 @@
     "No account yet?": "No account yet?",
     "sign up now": "sign up now",
     "To access": "To access",
-    "Sign in with {type}": "Sign in with {type}"
+    "Sign in with {type}": "Sign in with {type}",
+    "Logged in successfully": "Logged in successfully"
```
Comment (hsluoyz) [inline web/src/locales/en/data.json:138]: Signed in

File: web/src/locales/zh/data.json [hunk 6]
```diff
@@ -129,7 +129,8 @@
     "No account yet?": "没有账号？",
     "sign up now": "立即注册",
     "To access": "访问",
-    "Sign in with {type}": "{type}登录"
+    "Sign in with {type}": "{type}登录",
+    "Logged in successfully": "登录成功"
```
Comment (hsluoyz) [inline web/src/locales/zh/data.json:133]: Don't we have this before?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/locales/zh/data.json (hunk 6 above)
Comment (sh1luo) [reply web/src/locales/zh/data.json:133]: Yes. All "Logged in successfully" message did not use i18n.
$ C3B$~2ZE)B %BJK0 W@09

**Comment by sh1luo** [comment]
@hsluoyz  done, now demo link: http://testb.rescld.cn/

**Comment by sh1luo** [comment]
@hsluoyz  plz review

**Comment by sh1luo** [comment]
@hsluoyz plz review

**Comment by sh1luo** [comment]
@hsluoyz plz review

**Comment by sh1luo** [comment]
@hsluoyz This time I enhanced maintainability and code readability, plz review

demo link: http://testb.rescld.cn/

**Comment by sh1luo** [comment]
@hsluoyz plz review

**Comment by sh1luo** [comment]
@hsluoyz review plz

**Comment by hsluoyz** [comment]
@sh1luo if verification code is wrong, then button hangs.

image

**Comment by hsluoyz** [comment]
@sh1luo plz fix conflicts

**Comment by sh1luo** [comment]
> @sh1luo plz fix conflicts

I know, but I don’t know why there are so many commits. I need to solve this problem first.

**Comment by sh1luo** [comment]
@hsluoyz everything is ok, plz review.

Demo link: http://testb.rescld.cn/

**Comment by sh1luo** [comment]
@hsluoyz  Why this pr is closed, and where is it not good enough?

**Comment by yorkane** [comment]
Why this pr is rejected?
Any update?

**Comment by hsluoyz** [comment]
The frontend code quality is not good. @yorkane can you make a new PR?

**Pull request closed without merging**